Colab-TextGen-GPU.ipynb

```diff
@@ -0,0 +1,144 @@
+        "  is_cuda118 = '+cu118' in torver  # 2.1.0+cu118\n",
+        "  is_cuda117 = '+cu117' in torver  # 2.0.1+cu117\n",
+        "\n",
+        "  textgen_requirements = open('requirements.txt').read().splitlines()\n",
+        "  if is_cuda117:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu117').replace('+cu122', '+cu117').replace('torch2.1', 'torch2.0') for req in textgen_requirements]\n",
+        "  elif is_cuda118:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu118').replace('+cu122', '+cu118') for req in textgen_requirements]\n",
+        "  with open('temp_requirements.txt', 'w') as file:\n",
+        "      file.write('\\n'.join(textgen_requirements))\n",
+        "\n",
+        "  !pip install -r extensions/openai/requirements.txt --upgrade\n",
+        "  !pip install -r temp_requirements.txt --upgrade\n",
+        "\n",
+        "  print(\"\\033[1;32;1m\\n --> If you see a warning about \\\"previously imported packages\\\", just ignore it.\\033[0;37;0m\")\n",
+        "  print(\"\\033[1;32;1m\\n --> There is no need to restart the runtime.\\n\\033[0;37;0m\")\n",
+        "\n",
+        "  try:\n",
+        "    import flash_attn\n",
+        "  except:\n",
+        "    !pip uninstall -y flash_attn\n",
+        "\n",
+        "# Parameters\n",
+        "model_url = \"https://huggingface.co/TheBloke/MythoMax-L2-13B-GPTQ\" #@param {type:\"string\"}\n",
+        "branch = \"gptq-4bit-32g-actorder_True\" #@param {type:\"string\"}\n",
+        "command_line_flags = \"--n-gpu-layers 128 --load-in-4bit --use_double_quant\" #@param {type:\"string\"}\n",
+        "api = False #@param {type:\"boolean\"}\n",
+        "\n",
+        "if api:\n",
+        "  for param in ['--api', '--public-api']:\n",
+        "    if param not in command_line_flags:\n",
+        "      command_line_flags += f\" {param}\"\n",
+        "\n",
+        "model_url = model_url.strip()\n",
+        "if model_url != \"\":\n",
+        "    if not model_url.startswith('http'):\n",
+        "        model_url = 'https://huggingface.co/' + model_url\n",
+        "\n",
+        "    # Download the model\n",
+        "    url_parts = model_url.strip('/').strip().split('/')\n",
+        "    output_folder = f\"{url_parts[-2]}_{url_parts[-1]}\"\n",
+        "    branch = branch.strip('\"\\' ')\n",
+        "    if branch.strip() not in ['', 'main']:\n",
+        "        output_folder += f\"_{branch}\"\n",
+        "        !python download-model.py {model_url} --branch {branch}\n",
+        "    else:\n",
+        "        !python download-model.py {model_url}\n",
+        "else:\n",
+        "    output_folder = \"\"\n",
+        "\n",
+        "# Start the web UI\n",
+        "cmd = f\"python server.py --share\"\n",
+        "if output_folder != \"\":\n",
+        "    cmd += f\" --model {output_folder}\"\n",
+        "cmd += f\" {command_line_flags}\"\n",
+        "print(cmd)\n",
+        "!$cmd"
```

Comment:
**Optimization Issue**: The code for launching the web UI does not efficiently handle different CUDA versions or potential issues with previously installed packages. It also includes unnecessary repetition and lacks optimal exception handling for the installation process. <br> **Fix**: Refactor the code to dynamically determine CUDA versions and handle package installations more robustly. Introduce a function to encapsulate the conditional logic for CUDA version handling and package installation. This will make the code cleaner and more maintainable. <br> **Code Suggestion**: 
 ```
 if Path.cwd().name != 'text-generation-webui':
   print("Installing the webui...")

   !git clone https://github.com/oobabooga/text-generation-webui
   %cd text-generation-webui

   # Refactor for dynamic CUDA version handling
   torver = torch.__version__
   print(f"TORCH: {torver}")
   is_cuda118 = '+cu118' in torver
   is_cuda117 = '+cu117' in torver

   def handle_cuda_versions(torver):
       if '+cu117' in torver:
           # Adjust requirements for CUDA 117
       elif '+cu118' in torver:
           # Adjust requirements for CUDA 118

   handle_cuda_versions(torver)
 ```